Hindsight Experience Replay as a replay buffer

Juno-T opened the pull request:
- [x] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [x] algorithm implementation fix
    + [x] documentation modification
    + [x] new feature
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using ~~`make commit-checks`~~ `make check-codestyle`(**required**)
- [x] If applicable, I have mentioned the relevant/related issue(s)
- [x] If applicable, I have listed every items in this Pull Request below

Related thread: [pr#510](https://github.com/thu-ml/tianshou/pull/510)

Hi, I've implemented Hindsight Experience Replay (HER) as a replay buffer and would like to contribute to tianshou.
I saw the discussion in the mentioned [pr#510](https://github.com/thu-ml/tianshou/pull/510) and it seems to be inactive for quite a while so I decided to make a new, independent pull request.

## implementation
I implemented HER solely as a replay buffer. It is done by temporarily directly re-writing transitions storage (`self._meta`) during the `sample_indices()` call. The original transitions are cached and will be restored at the beginning of the next sampling or when other methods is called. This will make sure that. for example, n-step return calculation can be done without altering the policy.

There is also a problem with the original indices sampling. The sampled indices are not guaranteed to be from different episodes. So I decided to perform re-writing based on the episode. This guarantees that the sampled transitions from the same episode will have the same re-written goal. This also make the re-writing ratio calculation slightly differ from the paper, but it won't be too different if there are many episodes in the buffer.

In the current commit, HER replay buffer only support 'future' strategy and online sampling. This is the best of HER in term of performance and memory efficiency.

I also add a few more convenient replay buffers (`HERVectorReplayBuffer`, `HERReplayBufferManager`), test env (`MyGoalEnv`), gym wrapper (`TruncatedAsTerminated`), unit tests, and a simple example (examples/offline/fetch_her_ddpg.py).

## verification
I have added unit tests for almost everything I have implemented.
HER replay buffer was also tested using DDPG on [`FetchReach-v3` env](https://github.com/Farama-Foundation/Gymnasium-Robotics). I used default DDPG parameters from mujoco example and didn't tune anything further to get this good result! (train script: examples/offline/fetch_her_ddpg.py).


![Screen Shot 2022-10-02 at 19 22 53](https://user-images.githubusercontent.com/42699114/193454066-0dd0c65c-fd5f-4587-8912-b441d39de88a.png)

### Todo
- [ ] Compile doc with `make doc`

Trinkle23897 commented:
The ubuntu GPU test is run on gym==0.25.2. I'll give some feedback later today.

Juno-T commented:
I see. I'll fix it after the feedback then.

Trinkle23897 commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,180 @@
+        """
+        if indices.size == 0:
+            return
+        # Construct episode trajectories
```
I do think you should sort indices here. An example:
```
indices = [9, 7, 8, 6, 4, 5, 3, 1, 2, 0]
done = [0, 0, 1, 0, 0, 1, 0, 0, 1, 1]
```
after some processing with horizon=8, `unique_ep_open_indices = [7, 4, 1, 0]` and `unique_ep_close_indices=[ 3,  0, -1,  9]`, which makes no sense

Juno-T replied:
Previously I sort indices in `sample_indices` method (since it is the only one that calls `rewrite_transitions`) but yeah, I agree that sort indices should be inside `rewrite_transitions`.

fixed: https://github.com/thu-ml/tianshou/commit/63e65c087fa4c070861f12ba4d5fea130fc56740

Trinkle23897 replied:
My bad. I didn't see L92 previously, but it also has an issue: if the replay buffer start index is not 0 (i.e., overwrite previous data point in a circular buffer), the sort approach will cause some bugs at some point. It should start with x where x is the closest index that is greater or equal to start index.

Juno-T replied:
You are right. This case somehow slipped my mind.
In `rewrite_transitions`, sorting indices serves as a way to group indices from the same episode together and order them chronologically. So to apply with the circular property, I'm thinking of sorting the indices, then shifting them so that the oldest index is at the front.
 
Will fix this later maybe today or tomorrow.

Juno-T replied:
Updated & add some test : https://github.com/thu-ml/tianshou/pull/753/commits/6ee7d08f414456ce2274e8ca0e6714e5cbe420c7

Trinkle23897 commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,180 @@
+        #   episode indices that will be altered
+        her_ep_indices = np.random.choice(
+            len(unique_ep_open_indices),
+            size=int(len(unique_ep_open_indices) * self.future_p),
```
Could you please briefly explain why we need `self.future_p` here, instead of all indices in this episode?

Juno-T replied:
Yes, you are correct. HER only performs goal re-writing on some random episodes (the amount is depending on the ratio parameter `future_k`, or in here, probablity `future_p`).

Trinkle23897 commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,180 @@
+            return
+        self._meta[self._altered_indices] = self._original_meta
+        # Clean
+        del self._original_meta, self._altered_indices
```
no need to manually delete them because it has been overwritten in the next line?

Juno-T replied:
fixed: https://github.com/thu-ml/tianshou/pull/753/commits/63e65c087fa4c070861f12ba4d5fea130fc56740

Juno-T commented on README.md:
```diff
@@ -39,6 +39,7 @@
 - [Generalized Advantage Estimator (GAE)](https://arxiv.org/pdf/1506.02438.pdf)
 - [Posterior Sampling Reinforcement Learning (PSRL)](https://www.ece.uvic.ca/~bctill/papers/learning/Strens_2000.pdf)
 - [Intrinsic Curiosity Module (ICM)](https://arxiv.org/pdf/1705.05363.pdf)
+- [Hindsight Experience Replay (HER)](https://arxiv.org/pdf/1707.01495.pdf)
```
updated: https://github.com/thu-ml/tianshou/pull/753/commits/63e65c087fa4c070861f12ba4d5fea130fc56740

Juno-T commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,180 @@
+    :param compute_reward_fn: a function that takes 3 arguments: \
+        `acheived_goal`, `desired_goal`, `info` and returns the reward(s).
+        Note that the goal arguments can have extra batch_size dimension and in that \
+        case, the rewards of size batch_size should be returned
```
updated: https://github.com/thu-ml/tianshou/pull/753/commits/63e65c087fa4c070861f12ba4d5fea130fc56740

Juno-T replied:
Here I also remove the third argument, `info`. Previously this argument was manually set to empty dict `{}` during the sampling because its shape was complicated. Originally I have it there because `Fetch` env's [`compute_reward`](https://github.com/Farama-Foundation/Gymnasium-Robotics/blob/main/gymnasium_robotics/envs/fetch_env.py#L67) function also takes `info` argument, although it was never used.

Juno-T commented:
Btw, I tried verifying documentation page by using `make doc` but I kept getting version mismatch error. The doc test seems to be ok so I think it shouldn't be a problem.

The error, just in case:
```
WARNING: while setting up extension sphinx.addnodes: node class 'meta' is already registered, its visitors will be overridden
/home/tianshou/venv/lib/python3.8/site-packages/sphinx/jinja2glue.py:106: DeprecationWarning: 'contextfunction' is renamed to 'pass_context', the old name will be removed in Jinja 3.1.
  def warning(context: Dict, message: str, *args: Any, **kwargs: Any) -> str:

Exception occurred:
  File "/home/tianshou/venv/lib/python3.8/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (docutils 0.19 (/home/tianshou/venv/lib/python3.8/site-packages), Requirement.parse('docutils<0.17,>=0.12'), {'Sphinx'})
The full traceback has been saved in /tmp/sphinx-err-_u9sq96y.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make[1]: *** [Makefile:20: html] Error 2
make[1]: Leaving directory '/home/tianshou/docs'
make: *** [Makefile:47: doc] Error 2
```

Trinkle23897 commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,183 @@
+class HERReplayBuffer(ReplayBuffer):
+    """Implementation of Hindsight Experience Replay. arXiv:1707.01495.
+
+    `HERReplayBuffer` is to be used with goal-based environment where the \
+    observation is a dictionary with keys `observation`, `achieved_goal` and \
+    `desired_goal`. Currently support only HER's future strategy, online sampling.
```
```suggestion
    HERReplayBuffer is to be used with goal-based environment where the
    observation is a dictionary with keys ``observation``, ``achieved_goal`` and
    ``desired_goal``. Currently support only HER's future strategy, online sampling.
```

Juno-T replied:
updated: https://github.com/thu-ml/tianshou/pull/753/commits/7802451a0200bdf4e4c0399a204038d9dcf74925

Trinkle23897 commented:
> Btw, I tried verifying documentation page by using make doc but I kept getting version mismatch error. The doc test seems to be ok so I think it shouldn't be a problem.

Have you tried to reinstall deps by `pip install -e ".[dev]"`?

Juno-T commented on tianshou/data/buffer/her.py:
```diff
@@ -0,0 +1,183 @@
+    :param int future_k: the 'k' parameter introduced in the paper. In short, there \
+        will be at most k episodes that are re-written for every 1 unaltered episode \
+        during the sampling.
+
```
updated: https://github.com/thu-ml/tianshou/pull/753/commits/7802451a0200bdf4e4c0399a204038d9dcf74925

Juno-T commented:
I have tried `pip install -e ".[dev]"` and manually downgraded to `docutils==0.16`. Both didn't work.
I wonder if the latest versions (installed during [`make doc`](https://github.com/thu-ml/tianshou/blob/master/Makefile)) of sphinx and docutils aren't compatible.
Could you share your sphinx and docutils version please?

Trinkle23897 commented on tianshou/data/buffer/her.py:
```diff
@@ -105,7 +105,12 @@ def rewrite_transitions(self, indices: np.ndarray) -> None:
             return
+
+        # Sort indices keeping chronological order
+        mask = indices >= self._index
+        indices[~mask] += self.maxsize
```
`indices[indices < self._index] += self.maxsize`

nuance1979 commented on examples/offline/fetch_her_ddpg.py:
```diff
@@ -0,0 +1,265 @@
+def get_args():
+    # python3 fetch_her_ddpg.py --task FetchReach-v3 --seed 0 --horizon 50
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--task", type=str, default="FetchReach-v3")
```
Is it possible to apply HER to other tasks, e.g., HalfCheetah-v3, Ant-v3, so that we can compare performances? If not, can you run DDPG on FetchReach-v3 with `mujoco_ddpg.py` to get a baseline number? Then you can update `examples/mujoco/README.md` with both numbers, i.e., DDPG with and without HER.

nuance1979 commented on examples/offline/fetch_her_ddpg.py:
```diff
@@ -0,0 +1,265 @@
+    return env, train_envs, test_envs
+
+
+class DictStateNet(Net):
```
I'm a bit concerned about the fact that the user need to follow the example in this file closely in order to apply HER to any other task. Can you rewrite `DictStateNet` as a [decorator](https://en.wikipedia.org/wiki/Decorator_pattern) of `Net` so that it can be applied to any `Net`? The way I see it, this decorator should reside in [common.py](https://github.com/thu-ml/tianshou/blob/master/tianshou/utils/net/common.py) or [continous.py](https://github.com/thu-ml/tianshou/blob/master/tianshou/utils/net/continuous.py) as a utility function to wrap any `Net` which wants to work with `HERReplayBuffer`.

nuance1979 commented on examples/offline/fetch_her_ddpg.py:
```diff
@@ -0,0 +1,265 @@
+        return super().forward(self.preprocess_obs(obs), state, info)
+
+
+class GoalStateCritic(Critic):
```
Same for `GoalStateCritic`.

Juno-T commented:
Sorry for missing for a few weeks, I had to clear up some work at hand.

In the latest update, https://github.com/thu-ml/tianshou/pull/753/commits/72b4ab446e4e6bd4f717a6d2669c96f4ac501595, I have:
- Debuged HER
- Updated documentation
- Updated the code as suggested 
(https://github.com/thu-ml/tianshou/pull/753#discussion_r988572521)
- Moved HER training example to mujoco folder 
(https://github.com/thu-ml/tianshou/pull/753#discussion_r989543191)
- Refactored the "dict state" code into a decorator according to the suggestion. I put it in `utils/net/common.py` 
(https://github.com/thu-ml/tianshou/pull/753#discussion_r989572583)
- Performed a simple evalutation against vanilla DDPG and reported the result in `examples/mujoco/README.md`. I'm not sure where to put it since HER is not a learning algorithm but a techqnique, so I created and put it in the "Others" section. 
(https://github.com/thu-ml/tianshou/pull/753#discussion_r989546459)

nuance1979 commented:
LGTM. Great work, @Juno-T !

nuance1979 commented:
@Trinkle23897 I have fixed the test failures. Feel free to approve or suggest more changes.

Trinkle23897 reviewed (approved):
LGTM! Thanks for the great works!
